rgbd/src/server.cpp

```diff
@@ -1,34 +1,46 @@
 #include "rgbd/server.h"
+#include "rgbd/tools.h"
+
+#include <std_msgs/String.h>
+
+#include <functional>
 
 namespace rgbd {
 
 // ----------------------------------------------------------------------------------------
 
-Server::Server()
+Server::Server() : pub_hostname_thread_ptr_(nullptr)
 {
+    const std::string& hostname = get_hostname();
+    hostname_ = hostname;
 }
 
 // ----------------------------------------------------------------------------------------
 
 Server::~Server()
 {
+    nh_.shutdown();
+    if (pub_hostname_thread_ptr_)
+        pub_hostname_thread_ptr_->join();
 }
 
 // ----------------------------------------------------------------------------------------
 
 void Server::initialize(const std::string& name, RGBStorageType rgb_type, DepthStorageType depth_type, const float service_freq)
 {
-    server_rgbd_.initialize(name, rgb_type, depth_type, service_freq);
-    server_shm_.initialize(name);
+    name_= name;
+    server_rgbd_.initialize(name_, rgb_type, depth_type, service_freq);
+    server_shm_.initialize(name_);
```

Comment:
Why not create the the `pub_hostname_thread_` here? Because you didn't receive an image yet? In that case I could see the point but don't find it very elegant, moreover since you need the `name_` and `hostname_` member variables.

Reply:
Indeed. Because you don't want to start publishing as you have not send an image yet. Because the shared memory is only openend, when the first image is send.
You can start publishing right away, in that case the shared memory client will be trying to connect, while the shared memory has not been created yet. Also not very elegant. But it might have the benefit that the client is already in SHM mode. If you start publishing at the moment you are sending the first image. Because of the delay ROS has for connecting pubs and subs (0,3s), you might miss the first image.

Reply:
As discussed: at this as a comment in the code near the location where the thread is constructed.